src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -6063,6 +6112,50 @@ ReservationPtr MergeTreeData::balancedReservation(
     return reserved_space;
 }
 
+ColumnsDescription MergeTreeData::getObjectColumns(
+    const DataPartsVector & parts, const ColumnsDescription & storage_columns)
+{
+    return DB::getObjectColumns(
+        parts.begin(), parts.end(),
+        storage_columns, [](const auto & part) -> const auto & { return part->getColumns(); });
+}
+
+ColumnsDescription MergeTreeData::getObjectColumns(
+    boost::iterator_range<DataPartIteratorByStateAndInfo> range, const ColumnsDescription & storage_columns)
+{
+    return DB::getObjectColumns(
+        range.begin(), range.end(),
+        storage_columns, [](const auto & part) -> const auto & { return part->getColumns(); });
+}
+
+void MergeTreeData::resetObjectColumnsFromActiveParts(const DataPartsLock & /*lock*/)
+{
+    const auto & columns = getInMemoryMetadataPtr()->getColumns();
+    if (!hasObjectColumns(columns))
+        return;
+
+    auto range = getDataPartsStateRange(DataPartState::Active);
+    object_columns = getObjectColumns(range, columns);
+}
```

Comment:
`updateObjectColumnsFromActiveParts` is also ok. By distinction between "reset" and "update" I meant that in case of "reset" columns description is recreated from scratch and this operation is much heavier than "update"